api/client/create.go

```diff
@@ -94,20 +93,37 @@ func (cli *DockerCli) createContainer(config *runconfig.Config, hostConfig *runc
 		defer containerIDFile.Close()
 	}
 
+	repo, tag := parsers.ParseRepositoryTag(config.Image)
+	if tag == "" {
+		tag = tags.DEFAULTTAG
+	}
+	ref := registry.ParseReference(tag)
+	var trustedRef registry.Reference
+
+	if isTrusted(repo) && !ref.IsTrusted() {
+		var err error
+		trustedRef, err = cli.trustedReference(repo, ref)
+		if err != nil {
+			return nil, err
+		}
+		config.Image = trustedRef.ImageName(repo)
+	}
+
 	//create the container
 	stream, _, statusCode, err := cli.call("POST", "/containers/create?"+containerValues.Encode(), mergedConfig, nil)
 	//if image not found try to pull it
 	if statusCode == 404 && strings.Contains(err.Error(), config.Image) {
-		repo, tag := parsers.ParseRepositoryTag(config.Image)
-		if tag == "" {
-			tag = tags.DEFAULTTAG
-		}
-		fmt.Fprintf(cli.err, "Unable to find image '%s' locally\n", utils.ImageReference(repo, tag))
+		fmt.Fprintf(cli.err, "Unable to find image '%s' locally\n", ref.ImageName(repo))
 
 		// we don't want to write to stdout anything apart from container.ID
 		if err = cli.pullImageCustomOut(config.Image, cli.err); err != nil {
 			return nil, err
 		}
+		if trustedRef != nil && !ref.IsTrusted() {
```

Comment:
I need to add a comment here. This is adding a tag from the an untrusted reference `ref` to a trusted reference `trustedRef`, this would be the case where you asked for `ubuntu:latest` but we pulled `ubunut@sha256...`, then will add the tag for `ubuntu:latest` pointing at `ubunut@sha256...`. This could possibly be done as well as pull `ubuntu:latest@sha256...` but would require additional code to support in the daemon.